In [57]:
import pandas as pd

In [58]:
df=pd.read_csv("BF_train.csv")

BLACK FRIDAY SALES : SALES DURING BLACK FRIDAY OFFERS 

DATASET : 55,000 OBSERVATIONS BASED ON 12 DIMENSIONS

OBJECTIVE : 

TO PREDICT THE PURCHASE DONE BY AN INDIVIDUAL

TO PREDICT THE PRODUCT CATEGORY PURCHASED BY AN INDIVIDUAL

In [59]:
df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN      8370  
1                 6.0                14.0     15200  
2                 NaN                 NaN      1422  
3                14.0                 NaN      1057  
4                 NaN                 NaN      7969

In [60]:
df.shape

(550068, 12)

In [61]:
df.dtypes

User_ID                         int64
Product_ID                     object
Gender                         object
Age                            object
Occupation                      int64
City_Category                  object
Stay_In_Current_City_Years     object
Marital_Status                  int64
Product_Category_1              int64
Product_Category_2            float64
Product_Category_3            float64
Purchase                        int64
dtype: object

# We can see that few dimensions which are of object types need to be converted into numeric/continuous.
 
-> GENDER, AGE, CITY_CATEGORY , STAY_IN_CURRENT_CITY_YEARS

# DATA PREPARATION

In [63]:
dummy=pd.get_dummies(df["Gender"])
dummy.head()

F  M
0  1  0
1  1  0
2  1  0
3  1  0
4  0  1

In [64]:
dummy1=pd.get_dummies(df["City_Category"])
dummy1.head()

A  B  C
0  1  0  0
1  1  0  0
2  1  0  0
3  1  0  0
4  0  0  1

In [65]:
df=pd.concat([df,dummy],axis=1)
df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  F  M  
0                 NaN                 NaN      8370  1  0  
1                 6.0                14.0     15200  1  0  
2                 NaN                 NaN      1422  1  0  
3                14.0                 NaN      1057  1  0  
4                 NaN                 NaN      7969  0  1

In [66]:
df=pd.concat([df,dummy1],axis=1)
df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  F  M  A  B  C  
0                 NaN                 NaN      8370  1  0  1  0  0  
1                 6.0                14.0     15200  1  0  1  0  0  
2                 NaN                 NaN      1422  1  0  1  0  0  
3                14.0                 NaN      1057  1  0  1  0  0  
4                 NaN                 NaN      7969  0  1  0  0  1

In [67]:
df=df.drop(["Gender","City_Category"],axis=1)
df.head()

User_ID Product_ID   Age  Occupation Stay_In_Current_City_Years  \
0  1000001  P00069042  0-17          10                          2   
1  1000001  P00248942  0-17          10                          2   
2  1000001  P00087842  0-17          10                          2   
3  1000001  P00085442  0-17          10                          2   
4  1000002  P00285442   55+          16                         4+   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0               0                   3                 NaN                 NaN   
1               0                   1                 6.0                14.0   
2               0                  12                 NaN                 NaN   
3               0                  12                14.0                 NaN   
4               0                   8                 NaN                 NaN   

   Purchase  F  M  A  B  C  
0      8370  1  0  1  0  0  
1     15200  1  0  1  0  0  
2      1422  1  0  1  0  0  
3      1057  1  0  1  0  0  
4      7969  0  1  0  0  1

In [68]:
df.dtypes

User_ID                         int64
Product_ID                     object
Age                            object
Occupation                      int64
Stay_In_Current_City_Years     object
Marital_Status                  int64
Product_Category_1              int64
Product_Category_2            float64
Product_Category_3            float64
Purchase                        int64
F                               uint8
M                               uint8
A                               uint8
B                               uint8
C                               uint8
dtype: object

In [69]:
df["Age"].value_counts()

26-35    219587
36-45    110013
18-25     99660
46-50     45701
51-55     38501
55+       21504
0-17      15102
Name: Age, dtype: int64

In [70]:
dummy3=pd.get_dummies(df["Age"])
dummy3.head()

0-17  18-25  26-35  36-45  46-50  51-55  55+
0     1      0      0      0      0      0    0
1     1      0      0      0      0      0    0
2     1      0      0      0      0      0    0
3     1      0      0      0      0      0    0
4     0      0      0      0      0      0    1

In [71]:
df=pd.concat([df,dummy3],axis=1)

In [72]:
df=df.drop("Age",axis=1)

In [73]:
df.head()

User_ID Product_ID  Occupation Stay_In_Current_City_Years  Marital_Status  \
0  1000001  P00069042          10                          2               0   
1  1000001  P00248942          10                          2               0   
2  1000001  P00087842          10                          2               0   
3  1000001  P00085442          10                          2               0   
4  1000002  P00285442          16                         4+               0   

   Product_Category_1  Product_Category_2  Product_Category_3  Purchase  F  \
0                   3                 NaN                 NaN      8370  1   
1                   1                 6.0                14.0     15200  1   
2                  12                 NaN                 NaN      1422  1   
3                  12                14.0                 NaN      1057  1   
4                   8                 NaN                 NaN      7969  0   

   ...  A  B  C  0-17  18-25  26-35  36-45  46-50  51-55  55+  
0  ...  1  0  0     1      0      0      0      0      0    0  
1  ...  1  0  0     1      0      0      0      0      0    0  
2  ...  1  0  0     1      0      0      0      0      0    0  
3  ...  1  0  0     1      0      0      0      0      0    0  
4  ...  0  0  1     0      0      0      0      0      0    1  

[5 rows x 21 columns]

In [74]:
df["Stay_In_Current_City_Years"].value_counts()

1     193821
2     101838
3      95285
4+     84726
0      74398
Name: Stay_In_Current_City_Years, dtype: int64

In [75]:
df["Stay_In_Current_City_Years"]=df["Stay_In_Current_City_Years"].replace("4+",4)

In [76]:
df["Stay_In_Current_City_Years"].value_counts()

1    193821
2    101838
3     95285
4     84726
0     74398
Name: Stay_In_Current_City_Years, dtype: int64

In [77]:
df["Stay_In_Current_City_Years"]=pd.to_numeric(df["Stay_In_Current_City_Years"])

In [78]:
df.head()

User_ID Product_ID  Occupation  Stay_In_Current_City_Years  Marital_Status  \
0  1000001  P00069042          10                           2               0   
1  1000001  P00248942          10                           2               0   
2  1000001  P00087842          10                           2               0   
3  1000001  P00085442          10                           2               0   
4  1000002  P00285442          16                           4               0   

   Product_Category_1  Product_Category_2  Product_Category_3  Purchase  F  \
0                   3                 NaN                 NaN      8370  1   
1                   1                 6.0                14.0     15200  1   
2                  12                 NaN                 NaN      1422  1   
3                  12                14.0                 NaN      1057  1   
4                   8                 NaN                 NaN      7969  0   

   ...  A  B  C  0-17  18-25  26-35  36-45  46-50  51-55  55+  
0  ...  1  0  0     1      0      0      0      0      0    0  
1  ...  1  0  0     1      0      0      0      0      0    0  
2  ...  1  0  0     1      0      0      0      0      0    0  
3  ...  1  0  0     1      0      0      0      0      0    0  
4  ...  0  0  1     0      0      0      0      0      0    1  

[5 rows x 21 columns]

In [79]:
import matplotlib.pyplot as plt

In [80]:
df.describe()

User_ID     Occupation  Stay_In_Current_City_Years  \
count  5.500680e+05  550068.000000               550068.000000   
mean   1.003029e+06       8.076707                    1.858418   
std    1.727592e+03       6.522660                    1.289443   
min    1.000001e+06       0.000000                    0.000000   
25%    1.001516e+06       2.000000                    1.000000   
50%    1.003077e+06       7.000000                    2.000000   
75%    1.004478e+06      14.000000                    3.000000   
max    1.006040e+06      20.000000                    4.000000   

       Marital_Status  Product_Category_1  Product_Category_2  \
count   550068.000000       550068.000000       376430.000000   
mean         0.409653            5.404270            9.842329   
std          0.491770            3.936211            5.086590   
min          0.000000            1.000000            2.000000   
25%          0.000000            1.000000            5.000000   
50%          0.000000            5.000000            9.000000   
75%          1.000000            8.000000           15.000000   
max          1.000000           20.000000           18.000000   

       Product_Category_3       Purchase              F              M  \
count       166821.000000  550068.000000  550068.000000  550068.000000   
mean            12.668243    9263.968713       0.246895       0.753105   
std              4.125338    5023.065394       0.431205       0.431205   
min              3.000000      12.000000       0.000000       0.000000   
25%              9.000000    5823.000000       0.000000       1.000000   
50%             14.000000    8047.000000       0.000000       1.000000   
75%             16.000000   12054.000000       0.000000       1.000000   
max             18.000000   23961.000000       1.000000       1.000000   

                   A              B              C           0-17  \
count  550068.000000  550068.000000  550068.000000  550068.000000   
mean        0.268549       0.420263       0.311189       0.027455   
std         0.443205       0.493601       0.462980       0.163405   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       1.000000       1.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

               18-25          26-35          36-45          46-50  \
count  550068.000000  550068.000000  550068.000000  550068.000000   
mean        0.181178       0.399200       0.199999       0.083082   
std         0.385166       0.489734       0.400000       0.276007   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       1.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

               51-55            55+  
count  550068.000000  550068.000000  
mean        0.069993       0.039093  
std         0.255136       0.193817  
min         0.000000       0.000000  
25%         0.000000       0.000000  
50%         0.000000       0.000000  
75%         0.000000       0.000000  
max         1.000000       1.000000

In [81]:
df.columns

Index(['User_ID', 'Product_ID', 'Occupation', 'Stay_In_Current_City_Years',
       'Marital_Status', 'Product_Category_1', 'Product_Category_2',
       'Product_Category_3', 'Purchase', 'F', 'M', 'A', 'B', 'C', '0-17',
       '18-25', '26-35', '36-45', '46-50', '51-55', '55+'],
      dtype='object')

# DROPPING USER_ID, PRODUCT_ID AS THEY ARE IDENTIFIER VARIABLES AND THEY DO NOT HAVE ANY IMPACT ON THE PURCHASE PREDICTION 

In [82]:
df=df.drop(["User_ID","Product_ID"],axis=1)

In [83]:
df.columns

Index(['Occupation', 'Stay_In_Current_City_Years', 'Marital_Status',
       'Product_Category_1', 'Product_Category_2', 'Product_Category_3',
       'Purchase', 'F', 'M', 'A', 'B', 'C', '0-17', '18-25', '26-35', '36-45',
       '46-50', '51-55', '55+'],
      dtype='object')

# DATA CLEANING

In [84]:
df.isnull().sum()

Occupation                         0
Stay_In_Current_City_Years         0
Marital_Status                     0
Product_Category_1                 0
Product_Category_2            173638
Product_Category_3            383247
Purchase                           0
F                                  0
M                                  0
A                                  0
B                                  0
C                                  0
0-17                               0
18-25                              0
26-35                              0
36-45                              0
46-50                              0
51-55                              0
55+                                0
dtype: int64

Product_Category_2 AND Product_Category_3 have null values.Many machine learning models are not capable of handling null values.So there is a need for null value imputation.

MOST SIMPLEST TECHNIQUES FOR NULL VALUE IMPUTATION

-> IDENTIFYING THE PERCENTAGE OF NULL VALUES IN A COLUMN(DIMENSION) ,
    IF MORE THAN 50%, THAT DIMENSION CONTRIBUTES VERY LESS INFORMATION TO THE DATASET.
    SO DROP THAT COLUMN
    OTHERWISE , REPLACE NULL VALUE WITH MEANINGFUL VALUES

-> REPLACING NULL VALUES WITH 0

-> REPLACING NULL VALUES WITH MEAN/MEDIAN

# GENRALLY , FOR A PRODUCT CATEGORY , IT MAKES SENSE TO ASSIGN 0 TO NULL VALUES



# BUILT REGRESSION MODELS WITH THE ABOVE MENTIONED NULL VALUE IMPUTATION TECHNIQUES AND FOUND THAT REPLACING NULL VALUES WITH 0 GIVES THE BEST ACCURACY.

In [87]:
df["Product_Category_2"]=df["Product_Category_2"].fillna(0)

In [88]:
df["Product_Category_3"]=df["Product_Category_3"].fillna(0)

In [89]:
df.head()

Occupation  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0          10                           2               0                   3   
1          10                           2               0                   1   
2          10                           2               0                  12   
3          10                           2               0                  12   
4          16                           4               0                   8   

   Product_Category_2  Product_Category_3  Purchase  F  M  A  B  C  0-17  \
0                 0.0                 0.0      8370  1  0  1  0  0     1   
1                 6.0                14.0     15200  1  0  1  0  0     1   
2                 0.0                 0.0      1422  1  0  1  0  0     1   
3                14.0                 0.0      1057  1  0  1  0  0     1   
4                 0.0                 0.0      7969  0  1  0  0  1     0   

   18-25  26-35  36-45  46-50  51-55  55+  
0      0      0      0      0      0    0  
1      0      0      0      0      0    0  
2      0      0      0      0      0    0  
3      0      0      0      0      0    0  
4      0      0      0      0      0    1

In [90]:
df.isnull().sum()

Occupation                    0
Stay_In_Current_City_Years    0
Marital_Status                0
Product_Category_1            0
Product_Category_2            0
Product_Category_3            0
Purchase                      0
F                             0
M                             0
A                             0
B                             0
C                             0
0-17                          0
18-25                         0
26-35                         0
36-45                         0
46-50                         0
51-55                         0
55+                           0
dtype: int64

# MODELLING

In [91]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error,r2_score

In [92]:
x=df.drop("Purchase",axis=1)
y=df["Purchase"]

In [93]:
x.columns

Index(['Occupation', 'Stay_In_Current_City_Years', 'Marital_Status',
       'Product_Category_1', 'Product_Category_2', 'Product_Category_3', 'F',
       'M', 'A', 'B', 'C', '0-17', '18-25', '26-35', '36-45', '46-50', '51-55',
       '55+'],
      dtype='object')

In [94]:
y.shape

(550068,)

In [95]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=0)

In [96]:
lm=LinearRegression()

In [97]:
lm.fit(x_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [98]:
lm.intercept_

10348.219499236235

In [99]:
list(zip(x_train.columns,lm.coef_))

[('Occupation', 6.025153114139672),
 ('Stay_In_Current_City_Years', 12.246085162298819),
 ('Marital_Status', -65.13340446991464),
 ('Product_Category_1', -349.7264521432455),
 ('Product_Category_2', 12.70980203464927),
 ('Product_Category_3', 142.1732545134695),
 ('F', -234.7524898725249),
 ('M', 234.7524898725246),
 ('A', -271.5611913913869),
 ('B', -104.57365226442174),
 ('C', 376.1348436558096),
 ('0-17', -503.5481886033821),
 ('18-25', -178.2424697421913),
 ('26-35', -4.624341042431833),
 ('36-45', 87.97187266943844),
 ('46-50', 68.11260605270137),
 ('51-55', 369.0419828608361),
 ('55+', 161.2885378050296)]

In [100]:
y_train_pred=lm.predict(x_train)
y_train_pred

array([12405.93848558, 12524.66808819,  9254.83900848, ...,
        8846.57217576,  9443.52183981,  8641.07799095])

In [101]:
y_pred=lm.predict(x_test)

In [102]:
y_pred

array([10544.13497857,  6489.76050797, 11181.16986742, ...,
        9537.12799019, 11330.4565093 ,  9369.31501608])

In [103]:
mse=mean_squared_error(y_test,y_pred)
r2=r2_score(y_test,y_pred)
r2_train=r2_score(y_train,y_train_pred)

In [104]:
print("Mean squared error : ",mse)
print("Train_r2_score : ",r2_train)
print("Test_r2_score : ",r2)

Mean squared error :  21426627.36608349
Train_r2_score :  0.15260016337414595
Test_r2_score :  0.15289207214024592


In [105]:
from sklearn.metrics import mean_absolute_error


In [106]:
mae=mean_absolute_error(y_test,y_pred)
print("Mean Absolute error : ",mae)

Mean Absolute error :  3534.847567680434


In [107]:
import numpy as np

In [108]:
import statsmodels.api as sm

x_train_sm=x_train.astype(float)


x_train_sm=sm.add_constant(x_train_sm)

lm_1=sm.OLS(y_train.astype(float),x_train_sm).fit()



C:\Users\RAMASAMY\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [109]:
lm_1.params

const                         5236.448421
Occupation                       6.025153
Stay_In_Current_City_Years      12.246085
Marital_Status                 -65.133404
Product_Category_1            -349.726452
Product_Category_2              12.709802
Product_Category_3             142.173255
F                             2383.471721
M                             2852.976701
A                             1473.921616
B                             1640.909155
C                             2121.617651
0-17                           244.515872
18-25                          569.821590
26-35                          743.439719
36-45                          836.035933
46-50                          816.176666
51-55                         1117.106043
55+                            909.352598
dtype: float64

In [110]:
print(lm_1.summary())

                            OLS Regression Results                            
Dep. Variable:               Purchase   R-squared:                       0.153
Model:                            OLS   Adj. R-squared:                  0.153
Method:                 Least Squares   F-statistic:                     4622.
Date:                Tue, 06 Aug 2019   Prob (F-statistic):               0.00
Time:                        22:57:41   Log-Likelihood:            -3.7956e+06
No. Observations:              385047   AIC:                         7.591e+06
Df Residuals:                  385031   BIC:                         7.591e+06
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
const               

# FROM THE ABOVE RESULTS , WE CAN SEE THAT MSE IS VERY LARGE AND R2 SCORE SHOWS MODEL ACCURACY OF 15% WHICH IS VERY LESS.HENCE WE NEED TO TRY OTHER MODELS WHICH WILL IMPROVE R2 SCORE

# RIDGE REGRESSION

In [ ]:
#RIDGE REGRESSION

In [111]:
from sklearn.linear_model import Ridge,Lasso
from sklearn.model_selection import GridSearchCV

In [112]:
lambdas=np.linspace(1,100,100)

In [113]:
params={'alpha':lambdas}

In [114]:
model=Ridge(fit_intercept=True)

In [115]:
grid_search=GridSearchCV(model,param_grid=params,cv=10,scoring="neg_mean_absolute_error")

In [116]:
grid_search.fit(x_train,y_train)

GridSearchCV(cv=10, error_score='raise-deprecating',
             estimator=Ridge(alpha=1.0, copy_X=True, fit_intercept=True,
                             max_iter=None, normalize=False, random_state=None,
                             solver='auto', tol=0.001),
             iid='warn', n_jobs=None,
             param_grid={'alpha': array([  1.,   2.,   3.,   4.,   5.,   6.,   7.,   8.,   9.,  10.,  11.,
        12.,  13.,  14.,  15.,  16.,  17.,  18.,  19.,  20.,  21.,  22.,
        23.,  24.,  25.,  26.,  27.,  28.,  29.,  30.,  31.,  32.,...
        34.,  35.,  36.,  37.,  38.,  39.,  40.,  41.,  42.,  43.,  44.,
        45.,  46.,  47.,  48.,  49.,  50.,  51.,  52.,  53.,  54.,  55.,
        56.,  57.,  58.,  59.,  60.,  61.,  62.,  63.,  64.,  65.,  66.,
        67.,  68.,  69.,  70.,  71.,  72.,  73.,  74.,  75.,  76.,  77.,
        78.,  79.,  80.,  81.,  82.,  83.,  84.,  85.,  86.,  87.,  88.,
        89.,  90.,  91.,  92.,  93.,  94.,  95.,  96.,  97.,  98.,  99.,
       100.]

In [117]:
grid_search.best_estimator_

Ridge(alpha=100.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [118]:
grid_search.cv_results_

{'mean_fit_time': array([0.48999934, 0.25480061, 0.22759953, 0.23839979, 0.26360023,
        0.26760037, 0.26119747, 0.23639965, 0.23200035, 0.25400033,
        0.23879979, 0.24399979, 0.24625282, 0.23215199, 0.23120003,
        0.23199906, 0.22879915, 0.2311996 , 0.23319986, 0.23359978,
        0.23759937, 0.23479955, 0.23319988, 0.24519987, 0.23519998,
        0.23280029, 0.24160004, 0.2483994 , 0.24639986, 0.25810072,
        0.2601651 , 0.25399966, 0.24399962, 0.2379998 , 0.23760006,
        0.22600052, 0.24599867, 0.2308001 , 0.23159966, 0.23799949,
        0.2384002 , 0.24920084, 0.24440031, 0.23880029, 0.25720065,
        0.23400033, 0.25319972, 0.24599979, 0.24720047, 0.23199956,
        0.23759921, 0.23839989, 0.24959893, 0.25280018, 0.24799919,
        0.24759901, 0.24360034, 0.24640038, 0.2488009 , 0.24919918,
        0.2563997 , 0.25080004, 0.24199996, 0.25640001, 0.24639962,
        0.25079968, 0.23959954, 0.23639987, 0.25119998, 0.23999958,
        0.24320028, 0.23679976,

In [119]:
ridge_model=grid_search.best_estimator_

In [120]:
ridge_model.fit(x_train,y_train)

Ridge(alpha=100.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [122]:
list(zip(x_train.columns,ridge_model.coef_))

[('Occupation', 6.031214932114572),
 ('Stay_In_Current_City_Years', 12.253544237258035),
 ('Marital_Status', -64.55377053553971),
 ('Product_Category_1', -349.7122011772977),
 ('Product_Category_2', 12.713026844238748),
 ('Product_Category_3', 142.177130335984),
 ('F', -234.61543998374896),
 ('M', 234.61543998381129),
 ('A', -271.30520032421),
 ('B', -104.50034134700637),
 ('C', 375.80554167390784),
 ('0-17', -498.91503191051305),
 ('18-25', -178.19390616819354),
 ('26-35', -4.967399704134791),
 ('36-45', 87.54372737290326),
 ('46-50', 67.42708973289785),
 ('51-55', 367.21236598949855),
 ('55+', 159.89315468856066)]

In [125]:
ytrainpred_ridge=ridge_model.predict(x_train)

In [126]:
ytrainpred_ridge

array([12406.44100701, 12524.56294872,  9253.6490968 , ...,
        8846.42688575,  9444.31783024,  8639.47133965])

In [123]:
ypred_ridge=ridge_model.predict(x_test)

In [124]:
ypred_ridge

array([10543.75637062,  6488.21010898, 11180.59469994, ...,
        9536.7931091 , 11330.14347821,  9369.28346513])

In [127]:
mse=mean_squared_error(y_test,ypred_ridge)
r2=r2_score(y_test,ypred_ridge)
r2_train=r2_score(y_train,ytrainpred_ridge)

In [128]:
print("Mean squared error : ",mse)
print("Train_r2_score : ",r2_train)
print("Test_r2_score : ",r2)

Mean squared error :  21426644.064219255
Train_r2_score :  0.15260012623761032
Test_r2_score :  0.1528914119746052


# LASSO REGRESSION

In [129]:
#LASSO


In [130]:
lambdas=np.linspace(1,10,100)
model=Lasso(fit_intercept=True)
params={"alpha":lambdas}

In [131]:
grid_search1=GridSearchCV(model,param_grid=params,cv=10,scoring="neg_mean_absolute_error")

In [132]:
grid_search1.fit(x_train,y_train)

GridSearchCV(cv=10, error_score='raise-deprecating',
             estimator=Lasso(alpha=1.0, copy_X=True, fit_intercept=True,
                             max_iter=1000, normalize=False, positive=False,
                             precompute=False, random_state=None,
                             selection='cyclic', tol=0.0001, warm_start=False),
             iid='warn', n_jobs=None,
             param_grid={'alpha': array([ 1.        ,  1.09090909,  1.18181818,  1.27272727,  1.36363636,
        1.45454545,  1.5454...
        7.81818182,  7.90909091,  8.        ,  8.09090909,  8.18181818,
        8.27272727,  8.36363636,  8.45454545,  8.54545455,  8.63636364,
        8.72727273,  8.81818182,  8.90909091,  9.        ,  9.09090909,
        9.18181818,  9.27272727,  9.36363636,  9.45454545,  9.54545455,
        9.63636364,  9.72727273,  9.81818182,  9.90909091, 10.        ])},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='neg_mean_absolu

In [133]:
lasso_model=grid_search1.best_estimator_

In [134]:
lasso_model.fit(x_train,y_train)

Lasso(alpha=10.0, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [172]:
y_pred_lasso=lasso_model.predict(x_test)

In [173]:
list(zip(x_train.columns,lasso_model.coef_))

[('Occupation', 7.078529814600221),
 ('Stay_In_Current_City_Years', 7.179221726018655),
 ('Marital_Status', -0.0),
 ('Product_Category_1', -348.2157864563475),
 ('Product_Category_2', 12.816612057979835),
 ('Product_Category_3', 142.47092651712626),
 ('F', -418.2446725739206),
 ('M', 1.4968283508370915e-12),
 ('A', -135.95997791028208),
 ('B', -0.0),
 ('C', 452.5932976028868),
 ('0-17', -129.70163095893489),
 ('18-25', -124.32496857356466),
 ('26-35', -0.0),
 ('36-45', 23.955129560121108),
 ('46-50', 0.0),
 ('51-55', 194.29432461393594),
 ('55+', 0.0)]

In [177]:
mse=mean_squared_error(y_test,y_pred_lasso)
r2=r2_score(y_test,y_pred_lasso)


In [178]:
print("Mean squared error : ",mse)
print("r2_score : ",r2)

Mean squared error :  21436724.140955348
r2_score :  0.15249289321704773


# DECISION TREE REGRESSION

In [137]:
from sklearn.tree import DecisionTreeRegressor
DT = DecisionTreeRegressor(max_depth=15, min_samples_leaf=100)
DT.fit(x_train,y_train)


DecisionTreeRegressor(criterion='mse', max_depth=15, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=100,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')

In [138]:
ytrainpred_dt=DT.predict(x_train)

In [139]:
ytrainpred_dt

array([13554.71603053, 14781.16111111,  6920.70050761, ...,
        6196.42809365,  9309.90909091,  8406.36206897])

In [140]:
ypred_dt=DT.predict(x_test)

In [143]:
mse=mean_squared_error(y_test,ypred_dt)
r2=r2_score(y_test,ypred_dt)
r2_train=r2_score(y_train,ytrainpred_dt)

In [144]:
print("Mean squared error : ",mse)
print("Train_r2_score : ",r2_train)
print("Test_r2_score : ",r2)

Mean squared error :  8639148.244937938
Train_r2_score :  0.6639581480921937
Test_r2_score :  0.6584487682916056


# RANDOM FOREST REGRESSOR

In [160]:
from sklearn.ensemble import RandomForestRegressor

In [161]:
RF = RandomForestRegressor(max_depth=8, min_samples_leaf=150)

In [162]:
RF.fit(x_train,y_train)

C:\Users\RAMASAMY\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=8,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=150, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=10,
                      n_jobs=None, oob_score=False, random_state=None,
                      verbose=0, warm_start=False)

In [163]:
ytrainpred_RF=RF.predict(x_train)

In [164]:
ytrainpred_RF

array([13477.81705605, 15092.57570492,  6367.89138194, ...,
        6291.82366965, 10171.1180483 ,  8265.61912978])

In [165]:
ypred_RF=RF.predict(x_test)

In [166]:
ypred_RF

array([12863.21342983,  1306.96795966, 15904.00990483, ...,
       10563.67610013, 10867.12287206,  6684.71339988])

In [167]:
from math import sqrt

In [168]:
mse=mean_squared_error(y_test,ypred_RF)
rmse=sqrt(mean_squared_error(y_test,ypred_RF))
r2=r2_score(y_test,ypred_RF)
r2_train=r2_score(y_train,ytrainpred_RF)

In [169]:
print("Mean squared error : ",mse)
print("Train_r2_score : ",r2_train)
print("Test_r2_score : ",r2)
print("rmse : ",rmse)

Mean squared error :  8803099.368319618
Train_r2_score :  0.6510740674500857
Test_r2_score :  0.6519669130735521
rmse :  2967.0017472727613


# FROM THE ABOVE MODELS WE CAN COME TO A CONCLUSION THAT , LINEAR REGRESSION MODEL IS NOT SUITABLE FOR THIS PREDICTION PROBLEM.DECISION TREE AND RANDOM FOREST MODELS GIVES THE BETTER ACCURACY OF 65% WITH MINIMIZED MSE. FURTHER TUNING OF THESE MODELS WITH RESPECT TO MULTI COLLINEARITY MAY RESULT IN HIGHER MODEL ACCURACY